spec_util/meld.go

```diff
@@ -446,15 +462,113 @@ func meldStruct(dst, src *pb.Struct) error {
 		}
 	}
 
-	// ... but what if we end up with hundreds of keys?
-	// Probably that means the key is part of the data. Let's use an arbitrary
-	// threshold, and if we cross it, merge all the keys.
-	// TODO: add another field to Struct in in the IR specifying that this has occurred.
-	// (and if we can infer a data format on the keys, it could live at that level too.)
+	// Apply a heuristic for deciding when to convert structs to maps.
+	if structShouldBeMap(dst) {
+		structToMap(dst)
+	}
```

Comment:
I've been concerned about the cost of this entire method (and about the cost of re-hashing the structure once it is done!)  And the additional loop does not help.

A small improvement would be to track whether we have made any changes or not, and skip this heuristic check (and, ideally, the re-hashing) if we have not.

Reply:
Won't be addressed in this PR.